src/lib.rs

```diff
@@ -487,66 +502,77 @@ impl<K: Eq + Hash, V, S: BuildHasher> CLruCache<K, V, S> {
     }
 
     /// Returns the value corresponding to the most recently used item or `None` if the cache is empty.
     /// Like `peek`, `font` does not update the LRU list so the item's position will be unchanged.
     pub fn front_mut(&mut self) -> Option<(&K, &mut V)> {
         self.storage
             .front_mut()
-            .map(|CLruNode { key, value }| (&*key.0, value))
+            .map(|CLruNode { key, value, .. }| (&*key.0, value))
     }
 
     /// Returns the value corresponding to the least recently used item or `None` if the cache is empty.
     /// Like `peek`, `back` does not update the LRU list so the item's position will be unchanged.
     pub fn back(&self) -> Option<(&K, &V)> {
         self.storage
             .back()
-            .map(|CLruNode { key, value }| (&*key.0, value))
+            .map(|CLruNode { key, value, .. }| (&*key.0, value))
     }
 
     /// Returns the value corresponding to the least recently used item or `None` if the cache is empty.
     /// Like `peek`, `back` does not update the LRU list so the item's position will be unchanged.
     pub fn back_mut(&mut self) -> Option<(&K, &mut V)> {
         self.storage
             .back_mut()
-            .map(|CLruNode { key, value }| (&*key.0, value))
+            .map(|CLruNode { key, value, .. }| (&*key.0, value))
     }
 
     /// Puts a key-value pair into cache.
     /// If the key already exists in the cache, then it updates the key's value and returns the old value.
     /// Otherwise, `None` is returned.
-    pub fn put(&mut self, key: K, value: V) -> Option<V> {
-        match self.lookup.entry(Key(Rc::new(key))) {
-            Entry::Occupied(mut occ) => {
-                let old = self.storage.remove(*occ.get());
-                let (idx, _) = self
-                    .storage
-                    .push_front(CLruNode {
-                        key: Key(occ.key().0.clone()),
-                        value,
-                    })
-                    .unwrap();
-                occ.insert(idx);
-                old.map(|CLruNode { value, .. }| value)
-            }
-            Entry::Vacant(vac) => {
-                let mut obsolete_key = None;
-                if self.storage.is_full() {
-                    obsolete_key = self.storage.pop_back().map(|CLruNode { key, .. }| key);
-                }
-                if let Some((idx, _)) = self.storage.push_front(CLruNode {
-                    key: Key(vac.key().0.clone()),
-                    value,
-                }) {
-                    vac.insert(idx);
-                }
-                if let Some(obsolete_key) = obsolete_key {
-                    self.lookup.remove(&obsolete_key);
-                }
-                None
+    pub fn put_with_weight(
+        &mut self,
+        key: K,
+        value: V,
+        weight: NonZeroUsize,
+    ) -> Result<Option<V>, ()> {
+        let weight = weight.get();
+        if weight > self.capacity() {
```

Comment:
If capacity serves as an element count limits and a weight limit at the same time, there are cases where we reserve way too many elements. E.g. if an application uses memory consumtion in bytes as weight and wants to limit weight to 2 GB at an average element size of 100 MB, 2 billion elements are reserved instead of 20.